examples/lambda_transform.cpp

```
@@ -0,0 +1,67 @@
+#include <Arduino.h>
+
+#include "math.h"
+#include "sensesp_app.h"
+#include "sensors/analog_input.h"
+#include "signalk/signalk_output.h"
+#include "transforms/lambda_transform.h"
+
+ReactESP app([]() {
+  SetupSerialDebug(115200);
+
+  sensesp_app = new SensESPApp("sensesp-illum-example", "My WiFi SSID",
+                               "my_wifi_password", "skdev.lan", 80, NONE);
+
+  const char* sk_path = "indoor.illumination";
+  const char* analog_in_config_path = "/indoor_illumination/analog_in";
+
+  uint read_delay = 500;
+
+#ifdef ESP8266
+  uint8_t pin = A0;
+#elif defined(ESP32)
+  uint8_t pin = 32;
+#endif
+
+  float output_scale = 3.3;
+
+  // Use AnalogInput as an example sensor. Connect it e.g. to a photoresistor
+  // or a potentiometer with a voltage divider to get an illustrative test
+  // input.
+
+  auto* analog_input =
+      new AnalogInput(pin, read_delay, analog_in_config_path, output_scale);
+
+  // This is our transform function. The example is artificial; a log transform
+  // with configurable multiplier, base, and offset parameters.
+
+  auto log_function = [](float input, float multiplier, float base,
+                         float offset) -> float {
+    return multiplier * log(input) / log(base) + offset;
+  };
+
+  // This is an array of parameter information, providing the keys and
+  // descriptions required to display user-friendly values in the configuration
+  // interface.
+
+  const ParamInfo* log_lambda_arg_data = new ParamInfo[3]{
+      {"multiplier", "Multiplier"}, {"base", "Base"}, {"offset", "Offset"}};
+
+  // Here we create a new LambdaTransform objects. The template parameters
+  // (four floats in this example) correspond to the following types:
+  // 1. Output type of the transform function
+  // 2. Input type of the transoform function
+  // 3. Type of parameter 1
```

Typo: "transoform".

---

This is the constructor for a LambdaTransform using a function with 3 parameters:
```LambdaTransform(O (*function)(I input, T1 param1, T2 param2, T3 param3),
                  T1 param1, T2 param2, T3 param3, const ParamInfo* param_info,
                  String config_path = "")```
That looks like the first parameter is the value of the function (`O (*function)(I input, T1 param1, T2 param2, T3 param3)`), the second, third, and fourth parameters are the types of the three parameters to the function (`param1`, `param2`,and `param3`), the fifth parameter is the pointer to the ParamInfo struct, and the last parameter is the config path.

You said you wanted comments on the public interface - my comment is that this is confusing to me because the comments in the code don't seem to correspond to the actual code.

---

Ah. I added the Base parameter afterwards and apparently forgot to update the comments. Will fix.

---

The comment listing was referring to the template type parameters, and actually the call missed one. It just happened to work because float is the default type. Basically, that's the drawback with my implementation approach: you have all the template type parameters always defined by they just default to float. The alternative would've been genuinely variable-length template parameter list but that didn't work with the compiler version and flags in use.

Should I make the default type something obscure to nudge people to be more explicit?

---

I made the default type `bool`. This will make the compiler throw an error if non-bool parameters are used in the lambda function and the template types are not explicitly defined.

---

That seems like a good solution, as the type is not likely to be `bool`.